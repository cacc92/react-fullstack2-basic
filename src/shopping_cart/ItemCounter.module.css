.itemRow {
  display: flex;
  align-items: center;
  gap: 10px;
  margin-top: 10px;
}

.textItem {
  width: 150px;
}

.red {
  color: red;
}
