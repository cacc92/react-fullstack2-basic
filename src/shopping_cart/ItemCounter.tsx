// Se realiza rafc para completar

import React, { useState, type CSSProperties } from "react";

// import "./ItemCounter.css";
import styles from "./ItemCounter.module.css";

interface Props {
  name: string;
  quantity?: number;
}

const sectionExample: CSSProperties = {
  display: "flex",
  alignItems: "center",
  gap: 10,
  marginTop: 10,
};

export const ItemCounter = ({ name, quantity = 1 }: Props) => {
  const [count, setCount] = useState(quantity);

  const nameProduct: CSSProperties = {
    // width: 150,
    // color: quantity === 1 ? "red" : " black",
    color: count === 1 ? "red" : " black",
  };

  // const itemCounterObj = useState(10);

  const handleClick = () => {
    console.log(`Click en ${name}`);
  };

  const handleAdd = () => {
    setCount(count + 1);
  };

  const handleSubstract = () => {
    if (count === 1) return;
    setCount(count - 1);
  };

  return (
    // <section style={sectionExample}>
    <section className={styles.itemRow}>
      <span className={styles.textItem} style={nameProduct}>
        {name}
      </span>
      <button
        // onMouseEnter={() => {
        //   console.log(`Mouse enter ${name}`);
        // }}
        // onClick={() => {
        //   handleClick();
        // }}
        onClick={handleAdd}
      >
        +1
      </button>
      <span>{count}</span>
      <button onClick={handleSubstract}>-1</button>
    </section>
  );
};
