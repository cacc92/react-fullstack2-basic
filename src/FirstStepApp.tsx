import { ItemCounter } from "./shopping_cart/ItemCounter";

interface ItemCart {
  productName: string;
  quantity: number;
}

const itemsInCart: ItemCart[] = [
  { productName: "Nintendo Switch 2", quantity: 1 },
  //{ productName: "Nintendo Switch 2", quantity: 1 },
  { productName: "Xbox Series X", quantity: 2 },
  { productName: "Playstation 5 Pro", quantity: 3 },
];

export function FirstStepApp() {
  return (
    <>
      {/* <h1>Hola Mundo!</h1>
      <p>Esto es un parrafo!</p>
      <button>Click me</button>
      <div>
        <h2>Hola dentro de un DIV</h2>
      </div> */}
      <h1>Carrito de compras</h1>
      {itemsInCart.map(({ productName, quantity }) => (
        <ItemCounter key={productName} name={productName} quantity={quantity} />
      ))}

      {/* <ItemCounter name="Nintendo Switch 2" quantity={5} />
      <ItemCounter name="Xbox Series X" quantity={10} />
      <ItemCounter name="Playstation 5 Pro" quantity={2} /> */}
    </>
  );
}
