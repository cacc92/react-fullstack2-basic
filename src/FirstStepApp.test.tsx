import { render, screen } from "@testing-library/react";
import { test, describe, expect, vi, afterEach } from "vitest";
import { FirstStepApp } from "./FirstStepApp";
import { ItemCounter } from "./shopping_cart/ItemCounter";


const mockItemCounter = vi.fn((props:unknown)=>{
    return (<div
        data-testid="ItemCounter"
        name={props.name}
        quantity={props.quantity}
    >
    </div>)
});

vi.mock("./shopping_cart/ItemCounter", () => ({
    ItemCounter: (props: unknown) => mockItemCounter(props)
}));


// vi.mock("./shopping_cart/ItemCounter", () => ({
//     ItemCounter: (props: unknown) => <div
//         data-testid="ItemCounter"
//         name={props.name}
//         quantity={props.quantity}
//     ></div>
// }))

describe("FirstStepApp",() =>{

    afterEach(() => {
        vi.clearAllMocks();
    })

    test("Deberia coincidir el snapshot", () =>{
        const {container} = render(<FirstStepApp />);
        screen.debug();
        expect(container).toMatchSnapshot();
    });

    test("Deberian construirse todos los elementos item counter",()=>{
        render(<FirstStepApp />);
        const itemCounters = screen.getAllByTestId('ItemCounter')
        expect(itemCounters.length).toBe(3);
    })

    test("Deberia llamarse 3 veces la funcion de itemCounter", ()=>{
        render(<FirstStepApp />);
        expect(mockItemCounter).toHaveBeenCalledTimes(3);
        expect(mockItemCounter).toHaveBeenCalledWith({
            name: "Nintendo Switch 2", quantity: 1
        });
        expect(mockItemCounter).toHaveBeenCalledWith({
            name: "Xbox Series X", quantity: 2
        });
        expect(mockItemCounter).toHaveBeenCalledWith({
            name: "Playstation 5 Pro", quantity: 3
        });
    })

});