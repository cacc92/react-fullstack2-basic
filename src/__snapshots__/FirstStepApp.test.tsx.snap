// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FirstStepApp > Deberia coincidir el snapshot 1`] = `
<div>
  <h1>
    Carrito de compras
  </h1>
  <div
    data-testid="ItemCounter"
    name="Nintendo Switch 2"
    quantity="1"
  />
  <div
    data-testid="ItemCounter"
    name="Xbox Series X"
    quantity="2"
  />
  <div
    data-testid="ItemCounter"
    name="Playstation 5 Pro"
    quantity="3"
  />
</div>
`;
